File: Python/백준 알고리즘/1260.py
'''
문제
그래프를 DFS로 탐색한 결과와 BFS로 탐색한 결과를 출력하는 프로그램을 작성하시오. 단, 방문할 수 있는 정점이 여러 개인 경우에는 정점 번호가 작은 것을 먼저 방문하고, 더 이상 방문할 수 있는 점이 없는 경우 종료한다. 정점 번호는 1번부터 N번까지이다.
'''
# 노드의 개수 , 연결된 간선의 개수, 탐색을 시작할 정점의 번호를 준다
# 예제를 파악해보면
'''
4 5 1
1 2
1 3
1 4
2 4
3 4
이런식으로 예제가 들어왔는데, 대충 보면 1이라는 노드에 연결되어 있는 노드는 2,3,4 이고 2는 4, 3은 4이런식인것 같다
'''


def BFS(graph, start_node):
    need_visit = list()
    visited = list()
    need_visit.append(start_node)
    while need_visit:
        node = need_visit.pop(0)
        if node not in visited:
            visited.append(node)
            need_visit.extend(graph[node])
    return visited


def DFS(graph, start_node):
    visited, need_visit = list(), list()
    need_visit.append(start_node)

    while need_visit:
        node = need_visit.pop()
        if node not in visited:
            visited.append(node)
            need_visit.extend(graph[node])

    return visited


N, M, V = list(map(int, input().split()))
i = 0
keyed = 0
graph = [[] for _ in range(N+1)]
value_list = list()
new_graph = dict()
bfs_graph = dict()

while i < M:
    key, value = map(int, input().split())
    graph[key-1].append(value)
    graph[value-1].append(key)
    i += 1
for idx, value in enumerate(graph):
    value.sort(reverse=True)
    new_graph[idx+1] = value

print(new_graph)

answers = DFS(new_graph, V)
print(*answers, sep=' ')
for idx, value in enumerate(graph):
    value.sort()
    bfs_graph[idx+1] = value
answers = BFS(new_graph, V)
print(*answers, sep=' ')

# 아 알았다
'''
graph = dict()
graph['1'] = ['2']
graph['2'] = []
graph['3'] = ['4', '1']
graph['4'] = []
graph['5'] = ['4', '2']
이거봐 좀 이상하잖아
분명 5는 4,2 랑 연결되어있는데, 4에는 아무런게 없네?
'''

File: JS/백준 알고리즘/1260.js
// DFS와 BFS

// 입력값 받기
const testCase = require("fs")
  .readFileSync(process.platform == "linux" ? "/dev/stdin" : "예제.txt")
  .toString()
  .trim()
  .split("\n");

// 첫째줄 : 정점의 개수 N , 간선의 개수 M, 탐색을 시작할 정점의 번호 V
const [N, M, V] = testCase.shift().split(" ").map(Number);
// 두번째 줄 부터 ~ 마지막 : 간선이 연결된 두 정점
const arr = testCase.map((el) => el.split(" ").map(Number));

// Map 만들기
const grahp = new Map();
arr.forEach((el) => {
  [from, to] = el;
  if (!grahp.get(from)) {
    grahp.set(from, []);
  }

  if (!grahp.get(to)) {
    grahp.set(to, []);
  }

  grahp.get(from).push(to);
  grahp.get(to).push(from);
});

// console.log(grahp);
// dfs 만들기 : recusive 를 이용해서
/* 
sudo code;
1. node와 방문한 리스트를 받아온다
2. 해당 노드를 방문 리스트에 넣는다 (방문 가능해서 dfs를 돌리는것이기 때문)
3. 해당 노드와 연결된 리스트를 순회하면서, 만약 아직 방문하지 않았다면 재귀를 보낸다
4. 더이상 방문하지 않은 곳이 없다면 방문 리스트를 내보낸다
*/
const dfsRecusive = (node, visited = []) => {
  visited.push(node);
  grahp.get(node).map((el) => {
    if (!visited.includes(el)) {
      visited = dfsRecusive(el, visited);
    }
  });
  return visited;
};

// dfs 만들기 : stack 2개를  이용해서
const dfsStack = (node) => {
  visited = [];
  needVisit = [];
  needVisit.push(node);
  while (needVisit.length >= 1) {
    curNode = needVisit.pop();
    if (!visited.includes(curNode)) {
      visited.push(curNode);
      needVisit = needVisit.concat(grahp.get(curNode));
    }
  }
  return visited;
};

// bfs 만들기 : stack + queue를 이용해서
const bfsQueue = (node) => {
  visited = [];
  needVisit = [];
  needVisit.push(node);
  while (needVisit.length >= 1) {
    curNode = needVisit.shift();
    if (!visited.includes(curNode)) {
      visited.push(curNode);
      needVisit = needVisit.concat(grahp.get(curNode));
    }
  }
  return visited;
};

// console.log(...dfsStack(V));
grahp.forEach((el) => el.sort((a, b) => b - a));
console.log(dfsStack(V).join(" ").trim());
grahp.forEach((el) => el.sort((a, b) => a - b));
console.log(bfsQueue(V).join(" ").trim());
